File: epi_judge_python/lowest_common_ancestor_with_parent.py
import functools

from test_framework import generic_test
from test_framework.binary_tree_utils import must_find_node
from test_framework.test_failure import TestFailure
from test_framework.test_utils import enable_executor_hook


def lca(node0, node1):
    def depth(u):
        d = 0
        while u.parent != None:
            d += 1
            u = u.parent
        return d

    path0, path1 = depth(node0), depth(node1)
    diff = abs(path0 - path1)
    if path0 > path1:
        for i in range(diff):
            node0 = node0.parent
    else:
        for i in range(diff):
            node1 = node1.parent

    while not node0 is node1:
        node0 = node0.parent
        node1 = node1.parent
    return node0

@enable_executor_hook
def lca_wrapper(executor, tree, node0, node1):
    result = executor.run(
        functools.partial(lca, must_find_node(tree, node0),
                          must_find_node(tree, node1)))

    if result is None:
        raise TestFailure("Result can't be None")
    return result.data


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("lowest_common_ancestor_with_parent.py",
                                       'lowest_common_ancestor.tsv',
                                       lca_wrapper))

File: epi_judge_java/epi/LowestCommonAncestorWithParent.java
package epi;
import epi.test_framework.BinaryTreeUtils;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
import epi.test_framework.TestFailure;
import epi.test_framework.TimedExecutor;
public class LowestCommonAncestorWithParent {

  public static BinaryTree<Integer> LCA(BinaryTree<Integer> node0,
                                        BinaryTree<Integer> node1) {
    int d0 = 0, d1 = 0;
    BinaryTree<Integer> it0 = node0, it1 = node1;

    while(it0 != null){
      it0 = it0.parent;
      d0++;
    }

    while(it1 != null){
      it1 = it1.parent;
      d1++;
    }

    int diff = Math.abs(d1 - d0);
    if (d0 < d1){
      for (int i =0; i < diff; i++) node1 = node1.parent;
    } else {
      for (int i =0; i < diff; i++) node0 = node0.parent;
    }

    while (node0 != node1){
      node0 = node0.parent;
      node1 = node1.parent;
    }


    return node0;
  }
  @EpiTest(testDataFile = "lowest_common_ancestor.tsv")
  public static int lcaWrapper(TimedExecutor executor, BinaryTree<Integer> tree,
                               Integer key0, Integer key1) throws Exception {
    BinaryTree<Integer> node0 = BinaryTreeUtils.mustFindNode(tree, key0);
    BinaryTree<Integer> node1 = BinaryTreeUtils.mustFindNode(tree, key1);

    BinaryTree<Integer> result = executor.run(() -> LCA(node0, node1));

    if (result == null) {
      throw new TestFailure("Result can not be null");
    }
    return result.data;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "LowestCommonAncestorWithParent.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
